File: Python/traceback/subset.py
from typing import *

class Solution:
    #78.子集
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res,path = [],[]
        def traceback(nums, idx):
            res.append(path[:])
            for i in range(idx, len(nums)):
                path.append(nums[i])
                traceback(nums, i+1)
                path.pop()
        traceback(nums, 0)
        return res

    #90.子集II
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res,path = [],[]
        nums.sort()
        def traceback(nums, idx):
            res.append(path[:])
            for i in range(idx, len(nums)):
                if i > idx and nums[i] == nums[i-1]:
                    continue
                path.append(nums[i])
                traceback(nums, i+1)
                path.pop()
        traceback(nums,0)
        return res

s = Solution()
res = s.subsets([1,2,3])
print(res)

res = s.subsetsWithDup([1,2,2])
print(res)

File: Cpp/Traceback/subset.cpp
#include "subset.h"

void _subsets(vector<int>& nums, vector<vector<int>>& res, vector<int>& path, int idx){
    res.push_back(path);
    for(int i = idx; i < nums.size(); i++){
        path.push_back(nums[i]);
        _subsets(nums, res, path, i + 1);
        path.pop_back();
    }
}
vector<vector<int>> Subset::subsets(vector<int>& nums){
    vector<vector<int>> res;
    vector<int> path;
    _subsets(nums, res, path, 0);
    return res;
}

void _subsetsWithDup(vector<int>& nums, vector<vector<int>>& res, vector<int>& path, int idx){
    res.push_back(path);
    for(int i = idx; i < nums.size(); i++){
        if(i > idx && nums[i] == nums[i - 1]){
            continue;
        }
        path.push_back(nums[i]);
        _subsetsWithDup(nums, res, path, i + 1);
        path.pop_back();
    }
}
vector<vector<int>> Subset::subsetsWithDup(vector<int>& nums){
    vector<vector<int>> res;
    vector<int> path;
    sort(nums.begin(), nums.end());
    _subsetsWithDup(nums, res, path, 0);
    return res;
}